---
title: "regioneReloaded"
author: "Roberto Malinverni"
output: BiocStyle::html_document   #rmarkdown::html_vignette
fig_width: 10
fig_height: 10 
vignette: >
  %\VignetteIndexEntry{regioneReloaded}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
 fig.width=6, 
 fig.height=6,
 fig.align = "center"
)
```

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

# regioneReload

The meaningful interpretation of overlaps between binding profiles of multiple 
chromatin regulators is a major challenge in epigenomics. To address this, 
in 2015 we published `r BiocStyle::Biocpkg("regioneR")`, an R package that we developed for the statistical
assessment of the association between genomic regions sets. One limitation of regioneR is that it can only
perform association analyisis between two region sets at a time. Here, we now present 
**regioneReload**, an R package that is the natural evolution of `r BiocStyle::Biocpkg("regioneR")` and allows to  calculate the statistical association between multiple regions sets at the same time. To compare z-scores coming from multiple analysis, different strategies have been introduced to normalize the z-score and to improve the 
p-value calculations. regioneReload also improves `r BiocStyle::Biocpkg("regioneR")` by adding new randomization methods and different functions for obtaining publication-ready graphs. Taken together, regioneReload aims to be a novel and precious addition to NGS tools and for whole genome analysis.

## regioneR limitations

`r BiocStyle::Biocpkg("regioneR")` is an R package created to test the associations between genomic region sets. The core of `r BiocStyle::Biocpkg("regioneR")` is a permutation test framework specifically designed to work in a genomic environment. The two main results that can be obtained with this method can be summarized in two graphs. Figure 1A shows the association observed between the two region sets under study, highlighting the distance calculated in standard deviations from the random distribution. Figure 1B shows the local z-score: a narrow peak, as the one shown, indicates that the association is highly dependent on the exact location of the region. One the contrary, a flat profile would suggest a regional association.

<!-- This section is titled regioneR limitations but it does not address this point directly. 
Either highlight the limitation of pairwise comparisons or change the title? -->


(figure 1)

![regioneR basic Graphs](figures/regioner_grphs.png){width=120%}

## Normalized Z-Score

regioneReload aims to integrate the framework previously developed for `r BiocStyle::Biocpkg("regioneR")` 
to calculate associations of different region sets simultaneously. To compare
associations deriving from different region sets, we introduce the concept of 
**normalized z-score**, which allows not only to compare different association 
events but also to work with subsets of data and speed up the calculations.
To compare different z-scores is necessary to normalize them. The values of z-score of different test is not possible to compare.In particular z-score is directly associate to the square root of the  number of regions present in RS1.
If we empirically evaluate the value of normal z-score using different number of regions.
<!-- Explain what exactly the nZS corrects? I mean, describe what is the statistical/mathematical limitation that it aims to solve.
The dependency of the ZS value on the number of regions in the region sets tested? -->
The normalized Z-Score is calculated as follows:

nZS = ZS / $\sqrt{n}$


![local zscore  colculation (provisional)](figures/normal_ZScore.png){width=90%}

The normal z-score was stabilized using 30% of the sample, this can be used to reduce the calculation time.
<!-- Missing Figure -->
![local zscore  colculation (provisional)](figures/ntimes_stabilizations.png){width=90%}
Other crucial factor for calculate the normal z-score is the ntime, the suggestion after our tests is to use a minimum of 5000 permutation to achieve a reproducible results.

# Quick start

```{r setup}
library("regioneReloaded")
```
<!-- Consider moving this to the Crosswise Analysis section -->
The permutation test can be performed on a list multiple region sets by the function `crosswisePermTest()`.
This process is computing intensive and its calculation time depends on the 
number of cores called by the parameter `mc.cores` (see `r BiocStyle::Biocpkg("regioneR")`).
The result of the permutation test obtained by running the code below is included as a pre-computed example dataset
which can be loaded into the environment by running the command `data("cw_Alien_RaR")`.

`AlienGenome` and `AlienRSList` are described in the next section of the vignette. 

```{r load _data, eval=FALSE}
#NOT RUN
  
  set.seed(42)
  cw_Alien_ReG<-crosswisePermTest(Alist = AlienRSList_narrow,
                            sampling = FALSE,
                            mc.cores= 25,
                            ranFUN = "resampleGenome",
                            evFUN = "numOverlaps",
                            genome = AlienGenome,
                            ntimes= 1000
  )
#
```

```{r}

data("cw_Alien")


cw_Alien_ReG<-makeCrosswiseMatrix(cw_Alien_ReG, pvcut = 1)

plotCrosswiseMatrix(cw_Alien_ReG, matrix_type = "crosswise")

plotCrosswiseMatrix(cw_Alien_ReG, matrix_type = "correlation")
```
<!-- Briefly describe the plots? -->

## Alien Genome and Alien RegionSet List

Permutation test of regions set on a eukaryotic genome can take a long time. To show the different regionReloaded features, a fake genome called the 'AlienGenome' was created to overcome this problem. This genome consists of four chromosomes of different lengths ranging from 2Mb to 0.1Mb.

![Graphical descrition of the AlienGenome](figures/AlienGenome_scheme.png){width=100%}

```{r}

AlienGenome <-
  toGRanges(data.frame(
    chr = c("AlChr1", "AlChr2", "AlChr3", "AlChr4"),
    start = c(rep(1, 4)),
    end = c(2e6, 1e6, 5e5, 1e5)
  ))


```

Four basic region sets were created on this genome. The first three ("regA", "regB", "regC"), were created using the function in the regionR package `createRandomRegions`. Each Regions set consists of 100 regions with an average width of 100 bp (the widths of the regions will have a standard deviation of 50 bp). Being randomly originated, the three regions sets should show no associations with each other.


```{r}
gnm <- AlienGenome

nreg=100

regA <-
  createRandomRegions(
    nregions = nreg,
    length.mean = 100,
    length.sd = 50,
    non.overlapping = TRUE,
    genome = gnm
  )

regB <-
  createRandomRegions(
    nregions = nreg,
    length.mean =  100,
    length.sd = 50 ,
    non.overlapping = TRUE,
    genome = gnm
  )

regC <-
  createRandomRegions(
    nregions = nreg,
    length.mean = 100,
    length.sd = 50,
    non.overlapping = TRUE,
    genome = gnm
  )

```

To generate "artificial" associations we can use the `similarRegionSet()` function, which produces sets of random regions with some degree of similarity to the original. For each set of regions regA, regB and regC, we create a list of region sets that share a percentage (90% to 10%) of regions with the original set.

```{r}

vectorPerc <- seq(0.1, 0.9, 0.1)

RsetA <-
  similarRegionSet(
    GR = regA,
    name = "regA",
    genome = gnm,
    vectorPerc = vectorPerc
  )
RsetB <-
  similarRegionSet(
    GR = regB,
    name = "regB",
    genome = gnm,
    vectorPerc = vectorPerc
  )
RsetC <-
  similarRegionSet(
    GR = regC,
    name = "regC",
    genome = gnm,
    vectorPerc = vectorPerc
  )

```

Next we create a RegionSet of 100 regions which shares half the regions of regA and half of regB called regAB.  From this RegionSet we create similar region sets again using `similarRegionSet()`.

```{r}
vectorPerc2 <- seq(0.2, 0.8, 0.2)
regAB <- c(sample(regA, nreg / 2), sample(regB, nreg / 2))
RsetAB <-
  similarRegionSet(
    GR = regAB,
    name = "regAB",
    genome = gnm,
    vectorPerc = vectorPerc2
  )

```

As a negative count of, we create sets of random regions that contain no regions of regA, regB, regC, regAB or no regions of the preceding regABC.

```{r}
reg_no_A <-
  createRandomRegions(
    nregions = nreg,
    length.mean = 100,
    length.sd = 50,
    non.overlapping = TRUE,
    genome = subtractRegions(gnm, regA)
  )

reg_no_B <-
  createRandomRegions(
    nregions = nreg,
    length.mean = 100,
    length.sd = 50,
    non.overlapping = TRUE,
    genome = subtractRegions(gnm, regB)
  )

reg_no_C <-
  createRandomRegions(
    nregions = nreg,
    length.mean = 100,
    length.sd = 50,
    non.overlapping = TRUE,
    genome = subtractRegions(gnm, regC)
  )
reg_no_AB <-
  createRandomRegions(
    nregions = nreg,
    length.mean = 100,
    length.sd = 50,
    non.overlapping = TRUE,
    genome = subtractRegions(gnm, c(regA, regB))
  )

reg_no_ABC <-
  createRandomRegions(
    nregions = nreg,
    length.mean = 100,
    length.sd = 50,
    non.overlapping = TRUE,
    genome = subtractRegions(gnm, c(regA, regB, regC))
  )
```


```{r}
```



Finally, we store all the generated sets of regions in a list called `AlienRSList` which we can use as input for the `crosswisePermTest()` function.
```{r}
Rset_NO <- list(reg_no_A, reg_no_B, reg_no_C, reg_no_AB, reg_no_ABC)

names(Rset_NO) <- c("reg_no_A", "reg_no_B", "reg_no_C", "reg_no_AB", "reg_no_ABC")



AlienRSList_narrow <- c(RsetA, RsetB, RsetC, RsetAB, Rset_NO)

summary(AlienRSList_narrow)

```

## Multi Permutation Test 
### Crosswise Analysis

`crosswisePermTest()` is the core statistical function of the package. By Crosswise analysis we mean the calculation of permutation tests between all possible combinations of a regions set `Alist` to a second regions set `BList`. The result of `crosswisePermTest()`  is an S4 object of class `genoMatriXeR` that contains three slots: *parameteres*, *multiOverlaps* and *matrix*.

* gMXR_obj
  + @parameters
    + Alist
    + Blist
    + sampling
    + fraction
    + min_sampling
    + ranFUN
    + evFUN
    + universe
    + adj_pv_method
    + max_pv
    + nc
    + matOrder
    + ntimes
  + @multiOverlaps
    + one field for each comparison
  + @matrix
    + GMat
    + GMat_pv
    + GMat_corX
    + GMat_corY
    + FitRow
    + FitCol


#### Parameters

The slot "parameters" stores all the parameters used to obtain the genomicMatriXer object.

```{r }
data("cw_Alien")
print(cw_Alien_ReG@parameters)
```

#### multiOverlaps
<!-- The concept of Alist and Blist have not been introduced before this point, explain in the intro to the crosswiseAnalysis or in the example function call. -->
The "multiOverlaps" slot contains a list of data frames. For each element of the list of region sets "Alist", a data frame is stored with 11 columns and a number of rows equal to the elements in "Blist". The columns of the data frame contain the following information:

+ **order.id:**       order of comparison
+ **name:**           name of Blist elements 
+ **n_regionA:**       number of permuted Alist regions.
+ **n_regionB:**       number of regions in each regions set of Blist 
+ **z_score:**         z-score calculated by [regioneR::permTest()]
+ **p_value:**         p.value calculated by [regioneR::permTest()]
+ **n_hits:**          number of hits computed using the evaluation function on a regions set of Alist versus each regions set of Blists
+ **mean_perm_test:**  mean of hits computed using the evaluation function on the permutated region set
+ **sd_perm_test:**    standard deviation of hits computed using the evaluation function on the permutated region set
+ **norm_zscore:**     normalized z-scores 
+ **adj_p_value:**     adjusted p.value 


```{r}

print(cw_Alien_ReG)


```

#### matrix

When the genoMatriXeR object is created, the "matrix" slot will have a NULL value.
<!-- Explain here how to create the matrix, what is its purpose or where is this explained? -->

```{r}

str(cw_Alien_ReG@matrix)

```

### Evaluation functions
regioneReload can use all the evaluation strategies present in
`r BiocStyle::Biocpkg("regioneR")`.

### Randomization functions

regioneReload can use all the randomization strategies present in
`r BiocStyle::Biocpkg("regioneR")`.

(adding phrases for new approach of resampleRegions)

```{r, eval=FALSE}
#NOT RUN

set.seed(42)
cw_Alien_RaR <-  crosswisePermTest(
  Alist = AlienRSList_narrow,
  Blist = AlienRSList_narrow,
  sampling = FALSE,
  genome = AlienGenome,
  per.chromosome=TRUE,
  ranFUN = "randomizeRegions",
  evFUN = "numOverlaps",
  ntimes= 1000,
  mc.cores = 20
)


set.seed(42)
cw_Alien_ReR <-  crosswisePermTest(
  Alist = AlienRSList_narrow,
  Blist = AlienRSList_narrow,
  sampling = FALSE,
  genome = AlienGenome,
  per.chromosome=TRUE,
  ranFUN = "resampleRegions",
  evFUN = "numOverlaps",
  ntimes= 1000,
  mc.cores = 20
)

set.seed(42)
cw_Alien_ReG <-  crosswisePermTest(
  Alist = AlienRSList_narrow,
  Blist = AlienRSList_narrow,
  sampling = FALSE,
  genome = AlienGenome,
  per.chromosome=TRUE,
  ranFUN = "resampleGenome",
  evFUN = "numOverlaps",
  ntimes= 100,
  mc.cores = 20
)

#

```

Using the default option the matrix clusterization method will be automatically selected.

```{r}

cw_Alien_RaR <- makeCrosswiseMatrix(cw_Alien_RaR)

cw_Alien_ReG <- makeCrosswiseMatrix(cw_Alien_ReG)

cw_Alien_ReR <- makeCrosswiseMatrix(cw_Alien_ReR)

```

To compare the output of the matrices created with different randomization functions, 
we need to use the fixed order of clusterization for all matrixes.

```{r}

X<-rownames(cw_Alien_ReG@matrix$GMat)
Y<-colnames(cw_Alien_ReG@matrix$GMat)
ord<-list(X=X,Y=Y)

p_RaR <- plotCrosswiseMatrix(cw_Alien_RaR,matrix_type = "crosswise",ord_mat = ord)
p_ReG <- plotCrosswiseMatrix(cw_Alien_ReG,matrix_type = "crosswise" ,ord_mat = ord)
p_ReR <- plotCrosswiseMatrix(cw_Alien_ReR,matrix_type = "crosswise",ord_mat = ord)


plot(p_RaR)

ggpubr::ggarrange(p_ReG, p_ReR, 
          labels = c("ReG", "ReR"),
          ncol = 2, nrow = 1)

p_RaR <- plotCrosswiseMatrix(cw_Alien_RaR,matrix_type = "correlation",ord_mat = ord)
p_ReG <- plotCrosswiseMatrix(cw_Alien_ReG,matrix_type = "correlation" ,ord_mat = ord)
p_ReR <- plotCrosswiseMatrix(cw_Alien_ReR,matrix_type = "correlation",ord_mat = ord)

ggpubr::ggarrange(p_RaR, p_ReG,  p_ReR, 
          labels = c("RaR", "ReG","ReR"),
          ncol = 2, nrow = 2)

```

## Plot Single permutation test result

With the function `plotSinglePT()`, it is possible to plot the association between just two regions sest from a gMXR object.


```{r}

plotSinglePT(cw_Alien_ReG, RS1 = "regA", RS2 = "regA_05")
```

```{r}
p_sPT1 <- plotSinglePT(cw_Alien_ReG, RS1 = "regA", RS2 = "regC")

p_sPT2 <- plotSinglePT(cw_Alien_ReG, RS1 = "regA", RS2 = "reg_no_A")

ggpubr::ggarrange(p_sPT1, p_sPT2,
          ncol = 2, nrow = 1)
 
```

## Plot Dimensionally Reduction

Dimensionality reduction refers to a strategy capable of representing complex (high-density) data in a low-density space while retaining some meaningful properties of the original data. The plotCrosswiseDimRes function allows three of the most widely used algorithms (PCA, tRSNE, UMAP) to be applied to represent genoMatriXeR objects.

The data is clustered using a clustering algorithm that can be selected from ("hclust", "kmeans" or "pam" ), giving the possibility of importing an external clustering method.

```{r}

set.seed(42)
plotCrosswiseDimRed(cw_Alien_ReG, nc = 6, type="PCA")

```

The data is clustered using a clustering algorithm that can be selected from ("hclust", "kmeans" or "pam" ), giving the possibility of importing an external clustering method.

```{r}
set.seed(42)
p_cdr_hc <- plotCrosswiseDimRed(cw_Alien_ReG, nc = 6, type="PCA", clust_met = "hclust")

set.seed(42)
p_cdr_pam <- plotCrosswiseDimRed(cw_Alien_ReG, nc = 6, type="PCA", clust_met = "pam")

ggpubr::ggarrange(p_cdr_hc, p_cdr_pam,
          ncol = 2, nrow = 1)
```



```{r}

lsRegSet<-list(regA="regA",regB="regB",regC="regC")

set.seed(42)
plotCrosswiseDimRed(cw_Alien_ReG, nc = 6, type="PCA",listRS = lsRegSet)

set.seed(42)
plotCrosswiseDimRed(cw_Alien_ReG, nc = 6, type="PCA",listRS = lsRegSet,ellipse = TRUE, emphasize =TRUE)
```

```{r}

set.seed(67)
plotCrosswiseDimRed(cw_Alien_ReG, nc = 6, type="tSNE",listRS = lsRegSet,ellipse = TRUE, emphasize =TRUE)
```

```{r}
set.seed(67)
plotCrosswiseDimRed(cw_Alien_ReG, nc = 6, type="UMAP",listRS = lsRegSet,ellipse = TRUE, emphasize =TRUE)

```


# Multi Local Zscore

Descrizione

```{r, eval=FALSE }

#NOT RUN
mlz_Alien_ReG<-multiLocalZscore(A = AlienRSList$regA,
                 Blist = AlienRSList,
                 ranFUN = "resampleGenome",
                 evFUN = "numOverlaps",
                 window = 100,
                 step = 1,
                 max_pv =1,
                 genome = AlienGenome,
                 mc.cores = 25)
```

```{r, eval=FALSE }
str(mLZ_regA_ReG)
```
To create the matrix slot 

```{r, eval=FALSE }
mLZ_regA_ReG <- makeLZMatrix(mLZ_regA_ReG)

```
plotting
```{r , eval=FALSE }
plotLocalZScoreMatrix(mLZ_regA_ReG, maxVal = "max")
```

# Analysis with real data

To test how perform  `r BiocStyle::Biocpkg("regioneReload")` in a real scenario we download all transcrption factors present in the ENCODE (ref) project relative to HepG2 (261 ChIPseq experimets from human liver cancer). The complete GRList is possible to download at ("....").

## CrossWiseMatrix calculation
```{r, eval=FALSE }
# NOT run 

set.seed(42)
CW <-  crosswisePermTest(Alist = lGR,    # change name
                        Blist = lGR,
                        sampling = FALSE, 
                        genome = hg19,
                        ranFUN = "resampleGenome",
                        evFUN = "numOverlaps",
                        ntimes= 1000,
                        mc.cores = 25
                        )
```

```{r, eval = FALSE}
urlData<-url("https://github.com/RMalinverni/RegioneReloaded_data/blob/main/cw_Real.RData?raw=true")
load(urlData)


CW <- makeCrosswiseMatrix(CW)
```

```{r, eval=FALSE }
plotCrosswiseMatrix(CW, matrix.type = "crosswise" )
```

```{r, eval=FALSE }
plotCrosswiseMatrix(CW, matrix.type = "correlation")

```

```{r, eval=FALSE}
set.seed(42)
plotCrosswiseDimRed(CW,nc = 15,ellipse = TRUE,type = "UMAP")
```

```{r, eval=FALSE }
set.seed(42)
plotCrosswiseDimRed(CW,nc = 15,ellipse = TRUE,type = "PCA",
                    emphasize = TRUE,
                    listRS = list(RAD21=c("CTCF_ENCFF237OKO")))

set.seed(42, eval=FALSE )
plotCrosswiseDimRed(CW,nc = 15,ellipse = TRUE,type = "tSNE",
                    emphasize = TRUE,
                    listRS = list(RAD21=c("CTCF_ENCFF237OKO")))


set.seed(42)
plotCrosswiseDimRed(CW,nc = 15,ellipse = TRUE,type = "UMAP",
                    emphasize = TRUE,
                    listRS = list(CTCF=c("CTCF_ENCFF237OKO")))
```

## Multilocal Z-score Calculation


```{r, eval=FALSE}
mlZ_CTCF<-multiLocalZscore(A = toplGR$CTCF_ENCFF237OKO, Blist = toplGR,
                           sampling = TRUE, min_sampling = 5000,
                           ranFUN = "resampleGenome",
                           window = 1500,
                           step = 50, mc.cores=20)


```


```{r, eval=FALSE  }
mlZ_CTCF<-makeLZMatrix(mlZ_CTCF,normalize = TRUE,centralize =3)
```
```{r, eval=FALSE }
plotLocalZScoreMatrix(mlZ_CTCF,
                      maxVal = "max",
                      size_lab = 5,
                      main="CTCF_ENCFF237OKO")
```


# Session Info
```{r, sessionInfo}
sessionInfo()
```

